### ISSUE TITLE
CA1063: Remove IDisposable from the list of interfaces

### ISSUE THREAD

**Issue by lisberPontes** [dotnet/roslyn-analyzers#3941]
### Description

CA1063: Microsoft.Design : Remove IDisposable from the list of interfaces implemented by 'XXXXX and override the base class Dispose implementation instead.

Ok, so after reading the documentation Microsoft about this rule:

"Every unsealed type that declares and implements the IDisposable interface must provide its own protected virtual void Dispose(bool) method. Dispose() should call Dispose(true), and the finalizer should call Dispose(false). If you create an unsealed type that declares and implements the IDisposable interface, you must define Dispose(bool) and call it. For more information, see Clean up unmanaged resources (.NET guide) and Dispose pattern."

Is a basic set of rules that can and to some extent are easy to adopt among developers.

### Configuration

1) Define a base class MyBase that implements an interface such as IMyInterface;
2) IMyInterface implements IDisposable
3) Define a derived class MyDerive which inherits from MyBase class

4) Define a virtual implementation of Dispose() method in MyBase class
5) Do not override Dispose method on MyDerive class

### Regression?
No, it has been like this since day 1

### Other information

It shouldn't raise any CA1063 error for a scenario like the one described before, right ?

**Comment by Evangelink** [comment]
If I got things right, here is the scenario you are describing:

```csharp
public interface IMyInterface : IDisposable {}

public class MyBase : IMyInterface
{
    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool disposing)
    {
    }
}

public class MyDerive : MyBase
{
}
```

In this case I don't have any issue reported and it does make sense to me. But with the following code I manage to reproduce the `CA1063` and I do think the issue makes sense:

```csharp
public interface IMyInterface : IDisposable {}

public class MyBase : IMyInterface
{
    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool disposing)
    {
    }
}

public class MyDerive : MyBase, IDisposable 
{
}
``` 

Am I missing something?

**Comment by lisberPontes** [comment]
Yes, this 'false positive' can be seen here: [Stack overflow](https://stackoverflow.com/questions/8925925/code-analysis-ca1063-fires-when-deriving-from-idisposable-and-providing-implemen):

`public class Reusable : IDisposable {

  ~Reusable() {
    Dispose(false);
  }

  public void Dispose() {
    Dispose(true);
    GC.SuppressFinalize(this);
  }

  protected virtual void Dispose(Boolean disposing) {
    // ...
  }

  public void DoSomething() {
    // ...
  }

}

public interface IFunctionality : IDisposable {

  void DoSomething();

  void DoSomethingElse();

}

public class Functionality : Reusable, IFunctionality {

  public void DoSomethingElse() {
    // ...
  }

#if WORK_AROUND_CA1063
  // Removes CA1063
  protected override void Dispose(Boolean disposing) {
    base.Dispose(disposing);
  }
#endif

}`

**Comment by Evangelink** [comment]
Thanks for the pointer, the issue is a little different to what I understood from your comment and what I did posted earlier.

May I ask which version of the analyzer you are currently using? I have tried with v3.0.0 and latest beta and in both cases I don't have the issue reported.

**Comment by lisberPontes** [comment]
I'm using [legacy analysis](https://docs.microsoft.com/en-gb/visualstudio/code-quality/static-code-analysis-for-managed-code-overview?view=vs-2017) . And I've opened this same bug ([here](https://github.com/dotnet/runtime/issues/40196)). Thus, @danmosemsft told me "@lisberPontes thanks for the issue. Please can you reopen it in https://github.com/dotnet/roslyn-analyzers." 

And here we are :)

**Comment by Evangelink** [comment]
Oh I see! I don't want to say something wrong but AFAIK the legacy analysis is no longer really "supported". Besides as we don't have the issue on the new analyzer I think that nothing will be done.

Tagging @mavasani for a more official answer here.

**Comment by mavasani** [comment]
@Evangelink is correct. We are no longer taking fixes to legacy binary analysis. Please add a source suppression or disable the rule for legacy analysis.
